Add Strimzi KRaft migration support blog post (#424)
strimzi/strimzi.github.io: _posts/2024-03-20-strimzi-kraft-migration.md

---
layout: post
title:  "Migrate your Strimzi-operated cluster from ZooKeeper to KRaft"
date: 2024-03-20
author: paolo_patierno
---

> In the previous [blog post](TBD) we explored how ZooKeeper is used to store Kafka clusters metadata and what are the current limitations.

Reviewer: Maybe you can use the Apache in the first sentence?

```suggestion
In the previous [blog post](TBD) we explored how Apache ZooKeeper is used to store Apache Kafka clusters metadata and what are the current limitations.
```

We introduced the new KRaft protocol which allows to overcome such limitations and store metadata in Kafka itself, taking into account that the ZooKeeper support will be removed soon.
For this reason, there is the need to migrate existing clusters from using ZooKeeper to KRaft for storing metadata.
We also covered the overall procedure to do that.
Such procedure is not that simple because of the several configuration changes to be made on the cluster nodes, as well as the multiple times the nodes have to be rolled.

> But if your cluster is running on Kubernetes and it is Strimzi-operated, we got you covered!
> In this blog post, we are going to see how the Strimzi cluster operator provides a semi-automated process to run the migration easily, forgetting all the manual steps you should be doing to reach the same goal.

Reviewer 2: ```suggestion
The good news is that, if your cluster is powered by Strimzi, this process is much easier, and can be driven by simply annotating your Kafka resource.
You don't need to worry about updating configurations and restarting cluster nodes, as this is all automated.
```

Author: I used part of the suggestions mixing with my content.

<!--more-->

### Strimzi migration support

> The migration procedure is not that easy and it relies on updating brokers and controllers configurations multiple times together with rolling the involved nodes across several phases.

Reviewer 2: This seems a repetition of the introduction above.

There is a lot of manual intervention even related to verify the migration status by looking at the KRaft controller leader log or checking the `kafka.controller:type=KafkaController,name=ZkMigrationState` metric.
But if you are using the Strimzi operator to run your current ZooKeeper-based cluster on Kubernetes, the migration process is semi-automated.

> What you have to do is just updating the `strimzi.io/kraft` annotation on the `Kafka` cluster resource in order to allow the internal migration Finite State Machine (FSM) to move across the different phases.

Reviewer: The blog post seems user focused. It does not have any code details etc. So I would leave out also the _finite state machine_ from it and just say that the operator takes you through the different phases.

The operator is going to apply the needed configuration to the nodes and roll them.
It is also in charge of checking the migration status looking at the metrics for you to make sure it was done successfully.

> Before latest Strimzi 0.40.0 release, the `strimzi.io/kraft` annotation on the `Kafka` custom resource was used to define a ZooKeeper-based cluster, by using the `disabled` value, or a KRaft-based cluster, by using the `enabled` value.

Reviewer: Screw the past ... I would just say that the migration is supported from Strimzi 0.40.0 and not focus on how things worked before.

Author: Tbh I think it's useful making the point that the `kraft` annotation is not new and the current values are used to define ZooKeeper or KRaft based clusters. Then we used the same annotation for the migration with two additional values.

It is now possible to use two more values for the `strimzi.io/kraft` annotation: 

* `migration` to start the migration process;
* `rollback` to revert the migration itself;

> During migration or rollback, the `enabled` and `disabled` values are still used to finalize the procedure by ending with a cluster in KRaft mode or keeping the ZooKeeper mode.

Reviewer 2: These annotations are now used to trigger additional logic in the context of migration, not simply to discriminate between ZK-based or KRaft-based clusters, so I think we can say "still" used.

Author: I didn't get what you mean here. I am highlighting that enabled and disabled are anyway used to finalize the migration or rollback procedure.

Reviewer 2: I mean they trigger additional logic in the context of migration, so they are not the same as before. The suggestion was to just remove the word "still".

> One of the main prerequisites for the migration is about having the ZooKeeper-based cluster using the `KafkaNodePool`(s) to run the brokers.

Reviewer: The (s) is wierd in combination with the formatitng. I would work around it:

```suggestion
One of the main prerequisites for the migration is about having the ZooKeeper-based cluster using the `KafkaNodePool` resource(s) to run the brokers.
```

Is it worth mentioning also _not using BTO_ as prerequisite? And the Kafka version?

Author: Yeah good point.

> If that is not the case for you, please refer to the official Strimzi documentation [here](https://strimzi.io/docs/operators/latest/deploying#proc-migrating-clusters-node-pools-str).

Reviewer: Should we link to the node pools blog posts as well? They give some background to it.

> Following there is a snippet of a `Kafka` custom resource highlighting node pools enabled and using ZooKeeper for metadata.

Reviewer 2: ```suggestion
These are the annotations your Kafka resource should have before starting the migration:
```

Author: I used a mixed text together with your suggestion.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/kraft: disabled # states that the cluster is ZooKeeper-based
    strimzi.io/node-pools: enabled
# ...
```

The deployed cluster has severals pods running for the Kafka brokers and ZooKeeper nodes.

```shell
NAME                                          READY   STATUS    RESTARTS   AGE
my-cluster-entity-operator-555785d699-kksrs   2/2     Running   0          17m
my-cluster-kafka-0                            1/1     Running   0          18m
my-cluster-kafka-1                            1/1     Running   0          18m
my-cluster-kafka-2                            1/1     Running   0          18m
my-cluster-zookeeper-0                        1/1     Running   0          20m
my-cluster-zookeeper-1                        1/1     Running   0          20m
my-cluster-zookeeper-2                        1/1     Running   0          20m
strimzi-cluster-operator-7ddf57685d-245pd     1/1     Running   0          17m
```

#### From ZooKeeper to full KRaft based brokers

> The first step is to deploy a `KafkaNodePool` custom resource which hosts the KRaft controllers and then changing the `strimzi.io/kraft` annotation from `disabled` to `migration` to trigger the process.

Reviewer 2: ```suggestion
The first step is to deploy a `KafkaNodePool` custom resource which is used to provision the KRaft controllers and then changing the `strimzi.io/kraft` annotation from `disabled` to `migration` to trigger the process.
```

Following a snippet of a `KafkaNodePool` custom resource to be deployed for getting the KRaft controllers running.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: controller
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
    - controller
  storage:
    type: jbod
    volumes:
      - id: 0
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
```

The corresponding `Kafka` custom resource is then updated in regards to the `strimzi.io/kraft` set to `migration`. 

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:

>     strimzi.io/kraft: migration # updated from "disabled" to "migration" to trigger the migration process

Reviewer: Would `kubectl annotate` work better than this examples with the explanatory comment?

Author: I was wondering if using the command was going to be the same of the documentation with a lack of value for the blog post. Maybe having both?

Author: I added command and left YAML as well.

    strimzi.io/node-pools: enabled
# ...
```

> The annotation change triggers the Strimzi operator to run the following steps for you:

Reviewer 2: ```suggestion
The annotation change is detected by the Strimzi operator which runs the following steps:
```

Author: Tbh just removed "for you" but I like the current statement.

> * deploy the KRaft controllers, which are configured with all the connection details to ZooKeeper and with the migration flag enabled;
> * configure the running brokers with the connection details to the KRaft controller quorum, the migration flag enabled and roll them;
> * checking the status of the migration, on each reconciliation, and updating the `Kafka` custom resource to provide such status to the user through the `status.kafkaMetadataState` field;
> * when the migration is done, the brokers are reconfigured to be not connected to ZooKeeper anymore, with the migration flag disabled and they are rolled again.

Reviewer 2: ```suggestion
* deploy the KRaft controllers, which are configured with all the connection details to ZooKeeper and with the migration flag enabled
* configure the running brokers with the connection details to the KRaft controller quorum, the migration flag enabled and roll them
* on each reconciliation, check the status of the migration and update the `Kafka` custom resource to provide such status to the user through`status.kafkaMetadataState`
* when the migration is done, reconfigure brokers to disconnect from ZooKeeper, with the migration flag disabled, and roll them again
```

Author: Still I used part of the suggestions only.

As you can see, just applying one single annotation value has covered most of the phases in the migration procedure with the user doing actually nothing but following what the operator is running for them.

While the migration is going on, you can see the cluster changing the metadata state during the migration process from `ZooKeeper` to `KRaftPostMigration`, and then waiting for the user to finalize it.

```shell
NAME         DESIRED KAFKA REPLICAS   DESIRED ZK REPLICAS   READY   METADATA STATE   WARNINGS
my-cluster   3                        3                     True    ZooKeeper        
my-cluster   3                        3                     True    ZooKeeper        

> my-cluster   3                        3                     True    ZooKeeper        True

Reviewer 2: Should we explain what these warnings mean?

Author: I had this in mind but it could mean going into much details about specifying thet IBP and LMF are specified so you have to remove them and so on.

Reviewer 2: I think it's better to spend some words, otherwise the user will be confused.

Author: I added a few lines about it.

Reviewer 2: Thanks

my-cluster   3                        3                             ZooKeeper        True
my-cluster   3                        3                             ZooKeeper        
my-cluster   3                        3                             ZooKeeper        
my-cluster   3                        3                             ZooKeeper        
my-cluster   3                        3                     True    KRaftMigration   
my-cluster   3                        3                     True    KRaftDualWriting   
my-cluster   3                        3                     True    KRaftPostMigration   
my-cluster   3                        3                     True    KRaftPostMigration   True
```

In this status, the cluster is still working in "dual-write" mode and the user can validate that everything is working fine before deciding to finalize the migration or rolling back to use ZooKeeper.

#### Migration finalization

At this point, the user can finalize the migration by changing the `strimzi.io/kraft` annotation to `enabled`.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:

>     strimzi.io/kraft: enabled # updated from "migration" to "enabled" to finalize the migration process

Reviewer: Same as earlier

Author: Ditto as above.

    strimzi.io/node-pools: enabled
# ...
```

The operator will configure the KRaft controllers without the ZooKeeper connection details, with the migration flag disabled and roll them.

> When both brokers and controllers are working in KRaft mode, the operator will delete all the resources related to ZooKeeper as well.

Reviewer: Should you mention also deleting the `.spec.zookeeper` section from the CR?

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/kraft: disabled # updated from "rollback" to "disabled" to finalize the rollback to ZooKeeper
    strimzi.io/node-pools: enabled
# ...
```

Finally, the cluster metadata state comes back to be `ZooKeeper`.

```shell
NAME         DESIRED KAFKA REPLICAS   DESIRED ZK REPLICAS   READY   METADATA STATE   WARNINGS
...
my-cluster   3                        3                             KRaftDualWriting     
my-cluster   3                        3                             KRaftDualWriting     
my-cluster   3                        3                     True    ZooKeeper            True
```

### Conclusion

While the manual migration process could be considered long and complex if your cluster is running on bare-metal or virtual machines, it is very easy if the cluster is running on Kubernetes instead because of the Strimzi support for it.
The Strimzi operator provides you a semi-automated process which needs just a few manual steps to update an annotation to migrate your cluster.
Together with a GitOps approach you could easily migrate multiple cluster at the time.

> If it is about testing on a development environment or a production one, let us know your experience about migrating your cluster with Strimzi.

Reviewer: Sound strange. Maybe like this?

```suggestion
Let us know your experience with the migration of your cluster with Strimzi.
```

---
layout: post
title:  "Migrate your Strimzi-operated cluster from ZooKeeper to KRaft"
date: 2024-03-20
author: paolo_patierno
---

In the previous [blog post](TBD) we explored how Apache ZooKeeper is used to store Apache Kafka clusters metadata and what are the current limitations.
We introduced the new KRaft protocol which allows to overcome such limitations and store metadata in Kafka itself, taking into account that the ZooKeeper support will be removed soon.
For this reason, there is the need to migrate existing clusters from using ZooKeeper to KRaft for storing metadata.

> We also covered the overall procedure to do that.

Reviewer 3: This looks a bit odd to me
Maybe something like
`The overall procedure was also covered.`

or maybe connect it to the previous sentence

```
For this reason, there is the need to migrate existing clusters from using ZooKeeper to KRaft for storing metadata, where the overall procedure was also covered.
```

but maybe it's just me 🤷

<!--more-->

### Strimzi migration support

If you are using the Strimzi operator to run your current ZooKeeper-based cluster on Kubernetes, the migration process is semi-automated.
What you have to do is just updating the `strimzi.io/kraft` annotation on the `Kafka` cluster resource in order to allow the internal migration service to move across the different phases.
The operator is going to apply the needed configuration to the nodes and roll them.
It is also in charge of checking the migration status looking at the metrics for you to make sure it was done successfully.

Before latest Strimzi 0.40.0 release, the `strimzi.io/kraft` annotation on the `Kafka` custom resource was used to define a ZooKeeper-based cluster, by using the `disabled` value, or a KRaft-based cluster, by using the `enabled` value.

It is now possible to use two more values for the `strimzi.io/kraft` annotation: 

* `migration` to start the migration process;
* `rollback` to revert the migration itself;

During migration or rollback, the `enabled` and `disabled` values are used to finalize the procedure by ending with a cluster in KRaft mode or keeping the ZooKeeper mode.

> One of the main prerequisites for the migration is about having the ZooKeeper-based cluster using the `KafkaNodePool` resource(s) to run the brokers.

Reviewer 3: ```suggestion
One of the main prerequisites for the migration is having the ZooKeeper-based cluster using the `KafkaNodePool` resource(s) to run the brokers.
```

If that is not the case for you, please refer to the official Strimzi documentation [here](https://strimzi.io/docs/operators/latest/deploying#proc-migrating-clusters-node-pools-str) and this [series of blog posts](https://strimzi.io/blog/2023/08/14/kafka-node-pools-introduction/) focused on node pools.

Before starting the migration, the `Kafka` custom resource has to have the `strimzi.io/node-pools: enabled` annotation to use node pools together with the `strimzi.io/kraft: disabled` which defines a ZooKeeper-based cluster.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/kraft: disabled # states that the cluster is ZooKeeper-based
    strimzi.io/node-pools: enabled
# ...
```

> Using the Unidirectional Topic Operator and a Kafka version greater or equal 3.7.0 are important prerequisites as well.

Reviewer 3: Instead of having it in the text, should there be some lists of prerequisites, so the users that would like to try it will see everything needed on one place?

Author: It's a blog post, usually tables are useful in documentation imho. A blog post should more tell a story.

Reviewer 3: I didn't mean table, but just list of prerequisites:
```
Before starting the migration from ZK to KRaft, there are few prerequisites that are important to :

* use ZK-based cluster with `KafkaNodePool` for the brokers
* use Unidirectional Topic Operator
* have Kafka version greater or equal 3.7.0
```

but yeah, if you think it would not makes sense, then I'm fine with that.

Author: I just like the following flow because it easily allows to add links for the node pool migration and related blog posts.

my-cluster   3                        3                             ZooKeeper        True
my-cluster   3                        3                             ZooKeeper        
my-cluster   3                        3                             ZooKeeper        
my-cluster   3                        3                             ZooKeeper        
my-cluster   3                        3                     True    KRaftMigration   
my-cluster   3                        3                     True    KRaftDualWriting   
my-cluster   3                        3                     True    KRaftPostMigration   
my-cluster   3                        3                     True    KRaftPostMigration   True
```

In this status, the cluster is still working in "dual-write" mode and the user can validate that everything is working fine before deciding to finalize the migration or rolling back to use ZooKeeper.
If the `Kafka` custom resource has the `inter.broker.protocol.version` and `log.message.format.version` parameters set in the `spec.kafka.config` section, and because they are not supported in KRaft, the operator reports warnings into the status (see `WARNINGS` column).

> The user has to remove the parameters at the end of the process causing an additional rolling of the nodes.

Reviewer 3: I'm thinking if it would make sense to have something like this:
```
Once the metadata state is in `KRaftPostMigration` state, you have two options how you can continue:

* finish the migration -- link to the "Migration finalization" section
* roll back to ZooKeeper -- link to the "Rollback support" section
```

I think it will make it more clear to the reader that from this state there are two options. Because if you read it as it is, this sentence can make it a bit confusing `When the migration is not finalized yet, so when the KRaft controllers are still connected to ZooKeeper, it is still possible to rollback.` -> "what is the exact moment I can still roll back to ZK?"

And yeah I know that it will "duplicate" the `deciding to finalize the migration or rolling back to use ZooKeeper.` part of the sentence.

Author: Good idea!

---
layout: post
title:  "Migrate your Strimzi-operated cluster from ZooKeeper to KRaft"
date: 2024-03-20
author: paolo_patierno
---

In the previous [blog post](https://strimzi.io/blog/2024/03/21/kraft-migration/) we explored how Apache ZooKeeper is used to store Apache Kafka clusters metadata and what are the current limitations.

> We introduced the new KRaft protocol which allows to overcome such limitations and store metadata in Kafka itself, taking into account that the ZooKeeper support will be removed soon.

Reviewer 4: `We` -> `The Apache Kafka project` or ``The Apache Kafka community`

Author: As "We" I was referring to the previous blog post where the authors gave an introduction on the KRAft protocol. But I can see it could be misleading, I will change with your suggestion.

<!--more-->

### Strimzi migration support

If you are using the Strimzi operator to run your current ZooKeeper-based cluster on Kubernetes, the migration process is semi-automated.
What you have to do is just updating the `strimzi.io/kraft` annotation on the `Kafka` cluster resource in order to allow the internal migration service to move across the different phases.
The operator is going to apply the needed configuration to the nodes and roll them.
It is also in charge of checking the migration status looking at the metrics for you to make sure it was done successfully.

Before latest Strimzi 0.40.0 release, the `strimzi.io/kraft` annotation on the `Kafka` custom resource was used to define a ZooKeeper-based cluster, by using the `disabled` value, or a KRaft-based cluster, by using the `enabled` value.

It is now possible to use two more values for the `strimzi.io/kraft` annotation: 

* `migration` to start the migration process;
* `rollback` to revert the migration itself;

> During migration or rollback, the `enabled` and `disabled` values are used to finalize the procedure by ending with a cluster in KRaft mode or keeping the ZooKeeper mode.

Reviewer 4: Instead of `retaining` what about `reverting to`?

> During migration or rollback, the `enabled` and `disabled` values are used to finalize the procedure by ending with a cluster in KRaft mode or keeping the ZooKeeper mode.

Author: Maybe "reverting to" is better. I agree. In the end it's the final step of reverting to ZooKeeper during a migration.

One of the main prerequisites for the migration is having the ZooKeeper-based cluster using the `KafkaNodePool` resource(s) to run the brokers.
If that is not the case for you, please refer to the official Strimzi documentation [here](https://strimzi.io/docs/operators/latest/deploying#proc-migrating-clusters-node-pools-str) and this [series of blog posts](https://strimzi.io/blog/2023/08/14/kafka-node-pools-introduction/) focused on node pools.

Before starting the migration, the `Kafka` custom resource has to have the `strimzi.io/node-pools: enabled` annotation to use node pools together with the `strimzi.io/kraft: disabled` which defines a ZooKeeper-based cluster.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/kraft: disabled # states that the cluster is ZooKeeper-based
    strimzi.io/node-pools: enabled
# ...
```

> Using the Unidirectional Topic Operator and a Kafka version greater or equal 3.7.0 are important prerequisites as well.

Reviewer 4: I don't think migration requires the Unidirectional Topic Operator. I'm assuming you can migrate with you don't use the topic operator at all. I guess is if you are using the Topic operator you must first adopt the new unidirectional version.

> Using the Unidirectional Topic Operator and a Kafka version greater or equal 3.7.0 are important prerequisites as well.

Author: Yeah let's say it's not migration per se, it's the final state of being a KRaft cluster. So without ZooKeeper the BTO cannot work. I will rephrase it, thanks.

> Using the Unidirectional Topic Operator and a Kafka version greater or equal 3.7.0 are important prerequisites as well.

Author: So I think the current sentence works fine.

The deployed cluster has severals pods running for the Kafka brokers and ZooKeeper nodes.

```shell
NAME                                          READY   STATUS    RESTARTS   AGE
my-cluster-entity-operator-555785d699-kksrs   2/2     Running   0          17m
my-cluster-kafka-0                            1/1     Running   0          18m
my-cluster-kafka-1                            1/1     Running   0          18m
my-cluster-kafka-2                            1/1     Running   0          18m
my-cluster-zookeeper-0                        1/1     Running   0          20m
my-cluster-zookeeper-1                        1/1     Running   0          20m
my-cluster-zookeeper-2                        1/1     Running   0          20m
strimzi-cluster-operator-7ddf57685d-245pd     1/1     Running   0          17m
```

> #### From ZooKeeper to full KRaft based brokers

Reviewer 4: Do we need `fully`? It's not like brokers can be partially KRaft in the Strimzi migration process

> #### From ZooKeeper to full KRaft based brokers

Reviewer 4: My point is that "full-KRaft" is a term you invented, it's not a defined concept in Apache Kafka. So it can have a confusing meaning.

> #### From ZooKeeper to full KRaft based brokers

Author: Yeah, I removed it.

If the `Kafka` custom resource has the `inter.broker.protocol.version` and `log.message.format.version` parameters set in the `spec.kafka.config` section, and because they are not supported in KRaft, the operator reports warnings into the status (see `WARNINGS` column).
The user has to remove the parameters at the end of the process causing an additional rolling of the nodes.

Once the metadata state is in `KRaftPostMigration` state, you have two options how you can continue:

* [finalize the migration](#migration-finalization) to have a full KRaft cluster.
* [rollback the process](#rollback-to-zookeeper) to keep using ZooKeeper.

#### Migration finalization

At this point, the user can finalize the migration by changing the `strimzi.io/kraft` annotation to `enabled`.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/kraft: enabled # updated from "migration" to "enabled" to finalize the migration process
    strimzi.io/node-pools: enabled
# ...
```

You can do that by annotating the custom resource.

```shell
kubectl annotate kafka my-cluster strimzi.io/kraft="enabled" --overwrite
```

> The operator will configure the KRaft controllers without the ZooKeeper connection details, with the migration flag disabled and roll them.

Reviewer 5: ```suggestion
The operator will configure the KRaft controllers without the ZooKeeper connection details, with the migration flag disabled, and roll them.
```

When both brokers and controllers are working in KRaft mode, the operator will delete all the resources related to ZooKeeper as well.
At this point, the user can delete the `.spec.zookeeper` section from the `Kafka` custom resource because it is not needed anymore.

> During the finalization, the cluster metadata state moves to the final `KRaft`.

Reviewer 5: ```suggestion
During the finalization, the cluster metadata state moves to the final `KRaft` state.
```

```shell
NAME         DESIRED KAFKA REPLICAS   DESIRED ZK REPLICAS   READY   METADATA STATE   WARNINGS
...
my-cluster   3                        3                     True    KRaftPostMigration   True
my-cluster   3                        3                     True    PreKRaft             True
my-cluster   3                        3                     True    KRaft                True
my-cluster                                                  True    KRaft                True
my-cluster                                                  True    KRaft                
```

> As you can see, just a couple of changes on the `strimzi.io/kraft` annotation make the migration really straight forward.

Reviewer 5: ```suggestion
As you can see, just a couple of changes on the `strimzi.io/kraft` annotation make the migration really straightforward.
```

> #### Rollback to ZooKeeper

Reviewer 5: ```suggestion
#### Rolling back the migration to ZooKeeper
```

Reviewer 4: What about `Rolling back the cluster to ZooKeeper`?

Author: Yeah, I prefer Mickael suggestion.

> When the migration is not finalized yet, so when the KRaft controllers are still connected to ZooKeeper, it is still possible to rollback.

Reviewer 5: ```suggestion
If the migration process is still not finalized, with the KRaft controllers remaining connected to ZooKeeper, it is still possible to initiate a rollback.
```

> By applying the `rollback` value on the `strimzi.io/kraft` annotation, the operator reconfigures the brokers with the connection details to ZooKeeper again and roll them.

Reviewer 5: ```suggestion
By applying the `rollback` value on the `strimzi.io/kraft` annotation, the operator reconfigures the brokers with the connection details to ZooKeeper again and rolls them.
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/kraft: rollback # updated from "migration" to "rollback" to revert the migration process
    strimzi.io/node-pools: enabled
# ...
```  

You can do that by annotating the custom resource.

```shell
kubectl annotate kafka my-cluster strimzi.io/kraft="rollback" --overwrite
```

> Because of the rollback, the cluster metadata state moves back to `KRaftDualWriting`, waiting for the user to finalize the revert.

Reviewer 5: ```suggestion
Following the rollback, the cluster metadata state moves back to `KRaftDualWriting`, waiting for the user to finalize the process.
```

my-cluster   3                        3                     True    KRaftPostMigration   
my-cluster   3                        3                     True    KRaftPostMigration   True
my-cluster   3                        3                     True    KRaftPostMigration   True
my-cluster   3                        3                     True    KRaftDualWriting     
my-cluster   3                        3                     True    KRaftDualWriting     True
my-cluster   3                        3                             KRaftDualWriting     
```

After that, you have to delete the `KafkaNodePool` hosting the KRaft controllers and then apply the `disabled` value annotation.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/kraft: disabled # updated from "rollback" to "disabled" to finalize the rollback to ZooKeeper
    strimzi.io/node-pools: enabled
# ...
```

You can do that by annotating the custom resource.

```shell
kubectl annotate kafka my-cluster strimzi.io/kraft="disabled" --overwrite
```

The annotation change triggers the Strimzi operator to run the following steps for you:

> * automatically delete the `/controller` znode on ZooKeeper to allow the brokers to elect the new controller among them in place of the previous KRaft controller;
> * reconfigure brokers without connection details to the KRaft controller quorum anymore but using ZooKeeper again and roll them;

Reviewer 5: ```suggestion
* Automatically delete the `/controller` znode on ZooKeeper to enable brokers to elect a new controller from among themselves, replacing the previous KRaft controller.
* Reconfigure brokers to remove connection details to the KRaft controller quorum and revert to using ZooKeeper. Then, initiate a rolling update.
```

> Finally, the cluster metadata state comes back to be `ZooKeeper`.

Reviewer 5: ```suggestion
Finally, the cluster metadata state returns to `ZooKeeper`.
```

```shell
NAME         DESIRED KAFKA REPLICAS   DESIRED ZK REPLICAS   READY   METADATA STATE   WARNINGS
...
my-cluster   3                        3                             KRaftDualWriting     
my-cluster   3                        3                             KRaftDualWriting     
my-cluster   3                        3                     True    ZooKeeper            True
```

### Conclusion

> While the manual migration process could be considered long and complex if your cluster is running on bare-metal or virtual machines, it is very easy if the cluster is running on Kubernetes instead because of the Strimzi support for it.

Reviewer 5: ```suggestion
While the manual migration process could be considered long and complex if your cluster is running on bare-metal or virtual machines, it is very easy if the cluster is running on Kubernetes because of the Strimzi support for it.
```

> The Strimzi operator provides you a semi-automated process which needs just a few manual steps to update an annotation to migrate your cluster.

Reviewer 5: ```suggestion
The Strimzi operator provides a semi-automated process which needs just a few manual steps to update an annotation to migrate your cluster.
```

> Together with a GitOps approach you could easily migrate multiple cluster at the time.

Reviewer 5: ```suggestion
Together with a GitOps approach you could easily migrate multiple clusters at the time.
```

The Apache Kafka community introduced the new KRaft protocol which allows you to overcome these limitations and store metadata in Kafka itself, taking into account that the ZooKeeper support will be removed soon.
For this reason, there is a need to migrate existing clusters from using ZooKeeper to KRaft for storing metadata.
We also covered the overall Kafka metadata migration procedure.
The migration procedure can be done without downtime, but it is not that straightforward because it requires manual configurations and several rolling updates.
The good news is that if your cluster is powered by Strimzi, this process is much easier and can be accomplished simply by annotating your `Kafka` resource.
In this blog post, we are going to show you how the Strimzi Cluster Operator provides a semi-automated process to run the migration easily, so that you don't need to worry about updating configurations and restarting cluster nodes.

<!--more-->

### Strimzi's migration support

If you are using the Strimzi operator to run your current ZooKeeper-based cluster on Kubernetes, the migration process is semi-automated.

> All you need to do is update the `strimzi.io/kraft` annotation on the `Kafka` custom resource. 

Reviewer: This makes it sound like you update it just once to `enabled`.  Maybe you can say something like ...
```suggestion
The migration process is managed by updating the `strimzi.io/kraft` annotation on the `Kafka` custom resource.
```

This allows the internal migration service to transition smoothly across the different phases.
The operator will apply the necessary configuration to the nodes and then initiate a rolling update.
It is also in charge of checking the migration status by looking at the metrics to make sure it was done successfully.

Before the Strimzi 0.40.0 release, the `strimzi.io/kraft` annotation on the `Kafka` custom resource was used to define a ZooKeeper-based cluster, by using the `disabled` value, or a KRaft-based cluster, by using the `enabled` value.

> It is now possible to use two more values for the `strimzi.io/kraft` annotation: 

Reviewer: Maybe we can make it more clear that the old stuff did not changed as the previous para might sound like it did changed?

```suggestion
This is still used even in Strimzi 0.40.0.
But now it is possible to use two additional values for the `strimzi.io/kraft` annotation: 
```

Author: Good suggestion, thanks.

    strimzi.io/node-pools: enabled
# ...
```

Using the Unidirectional Topic Operator and a Kafka version greater or equal 3.7.0 are important prerequisites as well.

Before migrating, a deployed cluster generally has severals pods running for the Kafka brokers and ZooKeeper nodes.

```shell
NAME                                          READY   STATUS    RESTARTS   AGE
my-cluster-entity-operator-555785d699-kksrs   2/2     Running   0          17m
my-cluster-kafka-0                            1/1     Running   0          18m
my-cluster-kafka-1                            1/1     Running   0          18m
my-cluster-kafka-2                            1/1     Running   0          18m
my-cluster-zookeeper-0                        1/1     Running   0          20m
my-cluster-zookeeper-1                        1/1     Running   0          20m
my-cluster-zookeeper-2                        1/1     Running   0          20m
strimzi-cluster-operator-7ddf57685d-245pd     1/1     Running   0          17m
```

#### Migrating from ZooKeeper to KRaft-based brokers

The initial migration step involves deploying a `KafkaNodePool` custom resource to provision the KRaft controllers. 

> Then, you change the `strimzi.io/kraft` annotation from `disabled` to `migration` to initiate the process.

Reviewer: ```suggestion
Then, you change the `strimzi.io/kraft` annotation on the `Kafka` custom resource from `disabled` to `migration` to initiate the process.
```

Here's a snippet of the `KafkaNodePool` custom resource needed to deploy and activate the KRaft controllers:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: controller
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
    - controller
  storage:
    type: jbod
    volumes:
      - id: 0
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
```

The corresponding `Kafka` custom resource is then updated to set the `strimzi.io/kraft` annotation to `migration`:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/kraft: migration # updated from "disabled" to "migration" to trigger the migration process
    strimzi.io/node-pools: enabled
# ...
```

You can do that by annotating the custom resource.

```shell

> kubectl annotate kafka my-cluster strimzi.io/kraft="migration" --overwrite

Reviewer: I personally would change the order for the `kubectl annotate` command and the YAML example. But up to you.

Author: I like the current state, like showing how the resource should be and right after saying how to do that. I could change the order it's not a big deal but I should rephrase even the text around.

Reviewer: Well, as I said, your call. But having the annotate command first would allow you to remove the ugly comment that doesn't fit the screen without scrolling as the annotate command would tell the people what to look for in the YAML.

If you are using the Strimzi operator to run your current ZooKeeper-based cluster on Kubernetes, the migration process is semi-automated.
It is managed by updating the `strimzi.io/kraft` annotation on the `Kafka` custom resource.
This allows the internal migration service to transition smoothly across the different phases.
The operator will apply the necessary configuration to the nodes and then initiate a rolling update.
It is also in charge of checking the migration status by looking at the metrics to make sure it was done successfully.

Before the Strimzi 0.40.0 release, the `strimzi.io/kraft` annotation on the `Kafka` custom resource was used to define a ZooKeeper-based cluster, by using the `disabled` value, or a KRaft-based cluster, by using the `enabled` value.

This is still used even in Strimzi 0.40.0.
But now it is possible to use two additional values for the `strimzi.io/kraft` annotation:

* `migration` to start the migration process
* `rollback` to revert the migration itself

During migration or rollback, the `enabled` and `disabled` values are used to finalize the procedure by ending with a cluster in KRaft mode or reverting to the ZooKeeper mode.

> One of the main prerequisites for the migration is having the ZooKeeper-based cluster using the `KafkaNodePool` resource(s) to run the brokers.

Reviewer 6: Should we mention the prerequisites for Strimzi version >= 0.40.0?

Author: Wdym? We are already referring to Strimzi 0.40.0 as mentioned some lines before. So now I am adding prerequisites since this version (even because migration was not there before). Or I just didn't get what you meant :-)

Reviewer 6: Yes, I know you've referring to Strimzi 0.40.0, but you didn't say it's one of the prerequisites. Of course if you think that is fine, I'm also OK. Thanks.

Author: I changed the first sentence in the paragraph as ... `If you are using the Strimzi operator 0.40.0 release to run ..."
I can't say "greater than" because it's the only release we have out now with this feature. But it would be obvious until we'll remove migration in one year because ZooKeeper not supported anymore.

Reviewer 6: Sounds good to me. Thanks.

```

The annotation change triggers the Strimzi operator to run the following steps:

* Deploy the KRaft controllers with all the necessary connection details to ZooKeeper and the migration flag enabled.
* Configure the running brokers with the connection details to the KRaft controller quorum, with the migration flag enabled, and initiate a rolling update.
* During each reconciliation, monitor the migration status by checking the `kafka.controller:type=KafkaController,name=ZkMigrationState` metric. Update the `Kafka` custom resource to reflect this status through the `status.kafkaMetadataState` field.
* Once the migration is complete, reconfigure the brokers to disconnect from ZooKeeper, disable the migration flag, and initiate another rolling update.

As you can see, just applying one single annotation value has covered most of the phases in the migration procedure with the user doing practically nothing, just following what the operator is running for them.

While the migration is going on, you can see the cluster changing the metadata state during the migration process from `ZooKeeper` to `KRaftPostMigration`, and then waiting for the user to finalize it.

> ```shell

Reviewer 3: Maybe it would be good to have the command here as well? Just to show how the users can watch the state change (if you think it's a good idea, this applies to other places as well)?

Author: I will put the command just here, because the other places are snippet from the same output while the migration is running. I mean the end the user runs `kubectl get kafka my-cluster -n kafka -w` just here at the beginning (not every time).
